pkg/controller/metrics/metrics_with_dynamic_labels.go

```diff
@@ -0,0 +1,177 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/openshift/hive/apis/hive/v1/metricsconfig"
+	"github.com/openshift/hive/pkg/constants"
+)
+
+// metricsWithDynamicLabels encompasses metrics that can have optional labels.
+type metricsWithDynamicLabels interface {
+	ObserveMetricWithDynamicLabels(metav1.Object, log.FieldLogger, map[string]string, float64)
+	RegisterMetricWithDynamicLabels()
```

Comment:
nit: it seems redundant, and takes up extra space, to include `MetricWithDynamicLabels` in these func names, considering the structs already have that in their names. (Along similar lines, I think there's actually a go linting rule that complains if you have a package named `foo` containing objects/methods/etc. named `FooWhatever`.)